File: 程序员面试金典/4.3特定深度节点链表/listOfDepth.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
from collections import deque

class Solution:
    def listOfDepth(self, tree: TreeNode) -> List[ListNode]:
        if tree is None:
            return []
        queue = deque()
        queue.append((tree, 0))
        cur_level = -1
        ans = []
        layer = None
        while queue:
            node, level = queue.popleft()
            if level != cur_level:
                cur_level = level
                if layer is not None:
                    ans.append(layer)
                layer = ListNode(node.val)
                p = layer
            else:
                p.next = ListNode(node.val)
                p = p.next
            if node.left is not None:
                queue.append((node.left, level + 1))
            if node.right is not None:
                queue.append((node.right, level + 1))
        ans.append(layer)
        return ans

File: 程序员面试金典/4.3/listOfDepth.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

// 前序遍历
class Solution {
    private List<ListNode> firsts = new ArrayList<>();
    private List<ListNode> lasts = new ArrayList<>();

    public ListNode[] listOfDepth(TreeNode tree) {
        createLevelList(tree, 0);
        return firsts.toArray(new ListNode[0]);
    }

    private void createLevelList(TreeNode root, int level) {
        if (root == null)
            return;
        if (lasts.size() == level) {
            ListNode node = new ListNode(root.val);
            firsts.add(node);
            lasts.add(node);
        }
        else {
            ListNode node = lasts.get(level);
            node.next = new ListNode(root.val);
            lasts.set(level, node.next);
        }
        createLevelList(root.left, level + 1);
        createLevelList(root.right, level + 1);
    }

}
